### ISSUE TITLE
Wrong input shape for Conv2D

### ISSUE THREAD

**Issue by thunderhoser** [fversaci/WaveTF#7]
Hi,

Thanks for the great package!  I tried running the example on the README page (transform and anti-transform) in TensorFlow 2.2.0, and I get the following error:

------------------------------------------------------

Traceback (most recent call last):
  File "/home/ralager/ml4convection/ml4convection/stuff.py", line 772, in <module>
    t1 = w.call(t0)
  File "/home/ralager/anaconda3/lib/python3.7/site-packages/WaveTF-0.1-py3.7.egg/wavetf/_base_wavelets.py", line 109, in call
  File "/home/ralager/anaconda3/lib/python3.7/site-packages/WaveTF-0.1-py3.7.egg/wavetf/_haar_conv.py", line 169, in kernel_function
  File "/home/ralager/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py", line 1729, in conv1d_v2
    dilations=dilations)
  File "/home/ralager/anaconda3/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py", line 574, in new_func
    return func(*args, **kwargs)
  File "/home/ralager/anaconda3/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py", line 574, in new_func
    return func(*args, **kwargs)
  File "/home/ralager/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py", line 1663, in conv1d
    name=name)
  File "/home/ralager/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 937, in conv2d
    _ops.raise_from_not_ok_status(e, name)
  File "/home/ralager/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 6653, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.InvalidArgumentError: input must be 4-dimensional[32,1,900,200,1] [Op:Conv2D]

------------------------------------------------------

It seems like an extra dimension is being added to the tensor somewhere.  Do you have a solution for this?

Thanks,

Ryan

**Comment by fversaci** [comment]
Do you pass the tests with pytest?
```bash
$ pytest   # alternatively: python3 -m pytest
```

And what about running the example files?
```bash
$ python3 examples/transform_and_antitransform.py
```

**Comment by thunderhoser** [comment]
Hi Francesco,


Thanks for the quick reply.  I fail 8 of 32 unit tests, all with the same type of error message ("input must be 4-dimensional").  In all cases the input is 5-D even though 4-D was expected.  I also get the same error message when running transform_and_antitransform.py.

I *may* have found a solution via reshaping.  I'll create a pull request.


Thanks,

Ryan

**Comment by fversaci** [comment]
Ok, I've just tested a few different versions of Tensorflow:
- in Tensorflow 2.4 and 2.6 everything works correctly
- in Tensorflow 2.2.0 the tests fail

So it seems there's some API incompatibility kicking in at some point...
I might create a `tf2.2` branch with your solution (and narrow TF to ≥ 2.4 in the `master` docs).

**Comment by fversaci** [comment]
Hi Ryan,
I confirm it's a change in the tf.nn.conv1d behavior in TF>2.2 (it now accepts more than 4 dimensions).
Your fix #8  for TF 2.2 works for the 2D case, and it is still compatible with TF 2.4 and 2.6.

However, it needs to be integrated a bit (since it only handles 2D cases it doesn't pass the unit tests).
I'm accepting your PR and I will add another commit to complete it.

Thanks!

**Comment by thunderhoser** [comment]
Awesome, thank you so much!
